In [66]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error, mean_squared_error

In [42]:
df = pd.read_parquet('yellow_tripdata_2023-01.parquet')

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

## Q1. Downloading the data

Read the data for January. How many columns are there?

**19**

In [44]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0         14.30                   2.5         0.00  
1                    1.0         16.90                   2.5         0.00  
2                    1.0         34.90                   2.5         0.00  
3                    1.0         20.85                   0.0         1.25  
4                    1.0         19.68                   2.5         0.00

In [45]:
df['duration'] = ((df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds() / 60).round(2)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3066766 entries, 0 to 3066765
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

In [47]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  \
0                    1.0         14.30                   2.5         0.00   
1                    1.0         16.90                   2.5         0.00   
2                    1.0         34.90                   2.5         0.00   
3                    1.0         20.85                   0.0         1.25   
4                    1.0         19.68                   2.5         0.00   

   duration  
0      8.43  
1      6.32  
2     12.75  
3      9.62  
4     10.83

In [48]:
df.describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  3.066766e+06                     3066766                     3066766   
mean   1.730215e+00  2023-01-17 00:22:26.288164  2023-01-17 00:38:06.427874   
min    1.000000e+00         2008-12-31 23:01:42         2009-01-01 14:29:11   
25%    1.000000e+00  2023-01-09 16:21:57.250000         2023-01-09 16:37:06   
50%    2.000000e+00  2023-01-17 08:42:29.500000  2023-01-17 08:58:30.500000   
75%    2.000000e+00         2023-01-24 16:26:27         2023-01-24 16:42:49   
max    2.000000e+00         2023-02-01 00:56:53         2023-02-02 09:28:47   
std    4.438480e-01                         NaN                         NaN   

       passenger_count  trip_distance    RatecodeID  PULocationID  \
count     2.995023e+06   3.066766e+06  2.995023e+06  3.066766e+06   
mean      1.362532e+00   3.847342e+00  1.497440e+00  1.663980e+02   
min       0.000000e+00   0.000000e+00  1.000000e+00  1.000000e+00   
25%       1.000000e+00   1.060000e+00  1.000000e+00  1.320000e+02   
50%       1.000000e+00   1.800000e+00  1.000000e+00  1.620000e+02   
75%       1.000000e+00   3.330000e+00  1.000000e+00  2.340000e+02   
max       9.000000e+00   2.589281e+05  9.900000e+01  2.650000e+02   
std       8.961200e-01   2.495838e+02  6.474767e+00  6.424413e+01   

       DOLocationID  payment_type   fare_amount         extra       mta_tax  \
count  3.066766e+06  3.066766e+06  3.066766e+06  3.066766e+06  3.066766e+06   
mean   1.643926e+02  1.194483e+00  1.836707e+01  1.537842e+00  4.882900e-01   
min    1.000000e+00  0.000000e+00 -9.000000e+02 -7.500000e+00 -5.000000e-01   
25%    1.140000e+02  1.000000e+00  8.600000e+00  0.000000e+00  5.000000e-01   
50%    1.620000e+02  1.000000e+00  1.280000e+01  1.000000e+00  5.000000e-01   
75%    2.340000e+02  1.000000e+00  2.050000e+01  2.500000e+00  5.000000e-01   
max    2.650000e+02  4.000000e+00  1.160100e+03  1.250000e+01  5.316000e+01   
std    6.994368e+01  5.294762e-01  1.780782e+01  1.789592e+00  1.034641e-01   

         tip_amount  tolls_amount  improvement_surcharge  total_amount  \
count  3.066766e+06  3.066766e+06           3.066766e+06  3.066766e+06   
mean   3.367941e+00  5.184907e-01           9.820847e-01  2.702038e+01   
min   -9.622000e+01 -6.500000e+01          -1.000000e+00 -7.510000e+02   
25%    1.000000e+00  0.000000e+00           1.000000e+00  1.540000e+01   
50%    2.720000e+00  0.000000e+00           1.000000e+00  2.016000e+01   
75%    4.200000e+00  0.000000e+00           1.000000e+00  2.870000e+01   
max    3.808000e+02  1.969900e+02           1.000000e+00  1.169400e+03   
std    3.826759e+00  2.017579e+00           1.833529e-01  2.216359e+01   

       congestion_surcharge   airport_fee      duration  
count          2.995023e+06  2.995023e+06  3.066766e+06  
mean           2.274231e+00  1.074086e-01  1.566899e+01  
min           -2.500000e+00 -1.250000e+00 -2.920000e+01  
25%            2.500000e+00  0.000000e+00  7.120000e+00  
50%            2.500000e+00  0.000000e+00  1.152000e+01  
75%            2.500000e+00  0.000000e+00  1.830000e+01  
max            2.500000e+00  1.250000e+00  1.002918e+04  
std            7.718454e-01  3.556511e-01  4.259435e+01

## Q2. Computing duration

Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

**42.59**

In [49]:
original_len = len(df)
print(original_len)

3066766


In [50]:
df_filtered = df[(df['duration'] >= 1) & (df['duration'] <= 60)]

filtered_len = len(df_filtered)
print(filtered_len)

3009173


In [51]:
fraction = filtered_len / original_len
print(fraction)

0.9812202822125979


## Q3. Dropping outliers

Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

**98%**

In [55]:
df_filtered['PULocationID'] = df_filtered['PULocationID'].astype(str)
df_filtered['DOLocationID'] = df_filtered['DOLocationID'].astype(str)

C:\Users\isabe\AppData\Local\Temp\ipykernel_18944\798103828.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['PULocationID'] = df_filtered['PULocationID'].astype(str)
C:\Users\isabe\AppData\Local\Temp\ipykernel_18944\798103828.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['DOLocationID'] = df_filtered['DOLocationID'].astype(str)


In [56]:
dicts = df_filtered[['PULocationID', 'DOLocationID']].to_dict(orient='records')

In [57]:
dv = DictVectorizer()
X = dv.fit_transform(dicts)

print(X.shape[1])

515


## Q4. One-hot encoding

Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

- Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will label encode them)
- Fit a dictionary vectorizer
- Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

**515**

In [64]:
y = df_filtered['duration'].values

model = LinearRegression()
model.fit(X, y)

y_pred = model.predict(X)

In [67]:
rmse = np.sqrt(mean_squared_error(y, y_pred))
print(f'RMSE: {rmse:.2f}')

RMSE: 7.65


## Q5. Training a model

- Train a plain linear regression model with default parameters, where duration is the response variable
- Calculate the RMSE of the model on the training data

What's the RMSE on train?

**7.64**

In [68]:
df_val = pd.read_parquet('yellow_tripdata_2023-02.parquet')

In [69]:
df_val.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2913955 entries, 0 to 2913954
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

In [70]:
df_val['duration'] = ((df_val['tpep_dropoff_datetime'] - df_val['tpep_pickup_datetime']).dt.total_seconds() / 60).round(2)

df_val = df_val[(df_val['duration'] >= 1) & (df_val['duration'] <= 60)]

df_val['PULocationID'] = df_val['PULocationID'].astype(str)
df_val['DOLocationID'] = df_val['DOLocationID'].astype(str)

dicts_val = df_val[['PULocationID', 'DOLocationID']].to_dict(orient='records')
X_val = dv.transform(dicts_val)

y_val = df_val['duration'].values
y_pred_val = model.predict(X_val)

rmse_val = np.sqrt(mean_squared_error(y_val, y_pred_val))
print(f'Validation RMSE: {rmse_val:.2f}')

Validation RMSE: 7.81


## Q6. Evaluating the model

Now let's apply this model to the validation dataset (February 2023).

What's the RMSE on validation?

**7.81**